packages/interaction-tracking/src/InteractionTracking.js

```diff
@@ -0,0 +1,341 @@
+  interactions.add(interaction);
+  ((interactionsRef: any): InteractionsRef).current = interactions;
+
+  if (enableInteractionTrackingObserver) {
+    // Update before calling callback in case it schedules follow-up work.
+    interaction.__count = 1;
+
+    let caughtError;
+    let didCatch = false;
+    let returnValue;
+    const subscriber = ((subscriberRef: any): SubscriberRef).current;
+
+    if (subscriber !== null) {
+      try {
+        subscriber.onInteractionTracked(interaction);
+        subscriber.onWorkStarted(interactions, threadID);
+      } catch (error) {
+        didCatch = true;
+        caughtError = caughtError || error;
+      }
+    }
+
+    try {
+      returnValue = callback();
+    } catch (error) {
+      didCatch = true;
+      caughtError = caughtError || error;
+    }
+
+    ((interactionsRef: any): InteractionsRef).current = prevInteractions;
+
+    if (subscriber !== null) {
+      try {
+        subscriber.onWorkStopped(interactions, threadID);
+      } catch (error) {
+        didCatch = true;
+        caughtError = caughtError || error;
+      }
+    }
+
+    interaction.__count = ((interaction.__count: any): number) - 1;
+
+    // If no async work was scheduled for this interaction,
+    // Notify subscribers that it's completed.
+    if (subscriber !== null && ((interaction.__count: any): number) === 0) {
+      try {
+        subscriber.onInteractionScheduledWorkCompleted(interaction);
+      } catch (error) {
+        didCatch = true;
+        caughtError = caughtError || error;
+      }
+    }
+
+    if (didCatch) {
+      throw caughtError;
+    } else {
+      return returnValue;
+    }
+  } else {
+    try {
+      return callback();
+    } finally {
+      ((interactionsRef: any): InteractionsRef).current = prevInteractions;
+    }
+  }
+}
+
+export function unsubscribe(subscriber: Subscriber): void {
+  if (enableInteractionTracking && enableInteractionTrackingObserver) {
+    ((subscriberRef: any): SubscriberRef).current = null;
+  }
+}
+
+export function wrap(
+  callback: Function,
+  threadID: number = DEFAULT_THREAD_ID,
+): Function {
+  if (!enableInteractionTracking) {
+    return callback;
+  }
+
+  const wrappedInteractions = ((interactionsRef: any): InteractionsRef).current;
+
+  if (enableInteractionTrackingObserver) {
+    const subscriber = ((subscriberRef: any): SubscriberRef).current;
+    if (subscriber !== null) {
+      subscriber.onWorkScheduled(wrappedInteractions, threadID);
+    }
+
+    // Update the pending async work count for the current interactions.
+    // Update after calling subscribers in case of error.
+    wrappedInteractions.forEach(interaction => {
+      interaction.__count = ((interaction.__count: any): number) + 1;
+    });
+  }
+
+  const wrapped = (...args) => {
+    const prevInteractions = ((interactionsRef: any): InteractionsRef).current;
+    ((interactionsRef: any): InteractionsRef).current = wrappedInteractions;
```

Comment:
Almost always, this callback will be called at or near the top of the stack. But if it isn't, shouldn't this "stack/accumulate" interactions the same way `track` does? (If so, why not?) This is one of the reasons I assumed you would implement `wrap` on top of `track`.

Reply:
My above comment mentions this. The behavior of wrap/track are intentionally different. `track` is _additive_ (stacking on top of what the current interactions are). `wrap` temporarily _restores_ interactions at a previous point in time.